File: src/lc/cache.py
"""Persistent metadata cache for Codex rollout discovery.

This module is intentionally independent from transcript adapters and the TUI.
"""

from __future__ import annotations

from dataclasses import dataclass
import json
import os
from pathlib import Path


CACHE_VERSION = 1


@dataclass(frozen=True, slots=True)
class CodexCacheEntry:
    """The complete cache contract for one rollout file."""

    mtime_ns: int
    size: int
    cwd: str
    session_id: str
    title: str | None
    title_scanned: bool

    @classmethod
    def from_json(cls, value: object) -> CodexCacheEntry | None:
        if not isinstance(value, dict):
            return None
        try:
            mtime_ns = value["mtime_ns"]
            size = value["size"]
            cwd = value["cwd"]
            session_id = value["session_id"]
            title = value.get("title")
            title_scanned = value["title_scanned"]
        except KeyError:
            return None
        if not (isinstance(mtime_ns, int) and isinstance(size, int)
                and isinstance(cwd, str) and isinstance(session_id, str)
                and (title is None or isinstance(title, str))
                and isinstance(title_scanned, bool)):
            return None
        return cls(mtime_ns, size, cwd, session_id, title, title_scanned)

    def to_json(self) -> dict[str, object]:
        return {
            "mtime_ns": self.mtime_ns,
            "size": self.size,
            "cwd": self.cwd,
            "session_id": self.session_id,
            "title": self.title,
            "title_scanned": self.title_scanned,
        }

    def matches(self, stat_result: os.stat_result) -> bool:
        return self.mtime_ns == stat_result.st_mtime_ns and self.size == stat_result.st_size


class CodexCache:
    """Read unchanged entries and atomically persist meaningful batches.

    A single currently-active rollout is intentionally re-read instead of
    rewriting the entire cache on every agent turn.
    """

    def __init__(self, path: Path):
        self.path = path
        self._stored = self._load()
        self._next: dict[str, CodexCacheEntry] = {}
        self._changes = 0

    def _load(self) -> dict[str, CodexCacheEntry]:
        try:
            with open(self.path, encoding="utf-8") as fh:
                data = json.load(fh)
        except (OSError, json.JSONDecodeError):
            return {}
        if not isinstance(data, dict) or data.get("version") != CACHE_VERSION:
            return {}
        raw_entries = data.get("sessions")
        if not isinstance(raw_entries, dict):
            return {}
        return {
            path: entry
            for path, raw in raw_entries.items()
            if isinstance(path, str) and (entry := CodexCacheEntry.from_json(raw)) is not None
        }

    def get(self, path: str, stat_result: os.stat_result) -> CodexCacheEntry | None:
        entry = self._stored.get(path)
        if entry is not None and entry.matches(stat_result):
            return entry
        self._changes += 1
        return None

    def remember(self, path: str, entry: CodexCacheEntry, *, changed: bool = False) -> None:
        self._next[path] = entry
        if changed:
            self._changes += 1

    def flush(self) -> None:
        if self._stored and self._changes <= 1 and abs(len(self._next) - len(self._stored)) <= 1:
            return
        try:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            temporary = self.path.with_suffix(".tmp")
            with open(temporary, "w", encoding="utf-8") as fh:
                json.dump(
                    {"version": CACHE_VERSION,
                     "sessions": {path: entry.to_json() for path, entry in self._next.items()}},
                    fh,
                    separators=(",", ":"),
                )
            os.replace(temporary, self.path)
        except OSError:
            pass

File: internal/sessions/cache.go
package sessions

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"unicode/utf8"
)

// Persistent metadata cache for Codex rollout discovery. The format matches
// the Python implementation's (src/lc/cache.py) apart from the version.

// codexCacheVersion 2 invalidates titles scanned before lc read Codex's
// item_completed events, goal objectives, and subagent metadata.
const codexCacheVersion = 2

// codexEntry is the complete cache contract for one rollout file.
type codexEntry struct {
	MtimeNS      int64
	Size         int64
	Cwd          string
	SessionID    string
	Title        *string // nil when no title was found
	TitleScanned bool
}

func (e codexEntry) matches(fi os.FileInfo) bool {
	return e.MtimeNS == fi.ModTime().UnixNano() && e.Size == fi.Size()
}

// codexCache reads unchanged entries and atomically persists meaningful
// batches. A single currently-active rollout is intentionally re-read instead
// of rewriting the entire cache on every agent turn.
type codexCache struct {
	path    string
	stored  map[string]codexEntry
	next    map[string]codexEntry
	order   []string // insertion order of next, as Python dicts keep it
	changes int
}

func newCodexCache(path string) *codexCache {
	return &codexCache{path: path, stored: loadCodexCache(path), next: map[string]codexEntry{}}
}

func loadCodexCache(path string) map[string]codexEntry {
	raw, err := os.ReadFile(path)
	if err != nil || !utf8.Valid(raw) {
		return map[string]codexEntry{}
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var data any
	if dec.Decode(&data) != nil || dec.More() {
		return map[string]codexEntry{}
	}
	m, ok := data.(obj)
	if !ok || !isCacheVersion(m["version"]) {
		return map[string]codexEntry{}
	}
	sessions, ok := m["sessions"].(obj)
	if !ok {
		return map[string]codexEntry{}
	}
	out := make(map[string]codexEntry, len(sessions))
	for p, v := range sessions {
		if e, ok := codexEntryFromJSON(v); ok {
			out[p] = e
		}
	}
	return out
}

// isCacheVersion compares a decoded JSON number with codexCacheVersion
// numerically, so 2 and 2.0 both match.
func isCacheVersion(v any) bool {
	x, ok := v.(json.Number)
	if !ok {
		return false
	}
	f, err := x.Float64()
	return err == nil && f == codexCacheVersion
}

// pyInt accepts what Python's isinstance(v, int) accepts from json.load:
// integer literals and booleans.
func pyInt(v any) (int64, bool) {
	switch x := v.(type) {
	case json.Number:
		if strings.ContainsAny(string(x), ".eE") {
			return 0, false
		}
		n, err := x.Int64()
		return n, err == nil
	case bool:
		if x {
			return 1, true
		}
		return 0, true
	}
	return 0, false
}

func codexEntryFromJSON(v any) (codexEntry, bool) {
	m, ok := v.(obj)
	if !ok {
		return codexEntry{}, false
	}
	var e codexEntry
	for _, k := range [...]string{"mtime_ns", "size", "cwd", "session_id", "title_scanned"} {
		if _, ok := m[k]; !ok {
			return e, false
		}
	}
	var ok1, ok2, ok3, ok4, ok5 bool
	e.MtimeNS, ok1 = pyInt(m["mtime_ns"])
	e.Size, ok2 = pyInt(m["size"])
	e.Cwd, ok3 = m["cwd"].(string)
	e.SessionID, ok4 = m["session_id"].(string)
	e.TitleScanned, ok5 = m["title_scanned"].(bool)
	if !(ok1 && ok2 && ok3 && ok4 && ok5) {
		return e, false
	}
	switch t := m["title"].(type) {
	case nil:
	case string:
		e.Title = &t
	default:
		return e, false
	}
	return e, true
}

func (c *codexCache) get(path string, fi os.FileInfo) (codexEntry, bool) {
	if e, ok := c.stored[path]; ok && e.matches(fi) {
		return e, true
	}
	c.changes++
	return codexEntry{}, false
}

func (c *codexCache) remember(path string, e codexEntry, changed bool) {
	if _, ok := c.next[path]; !ok {
		c.order = append(c.order, path)
	}
	c.next[path] = e
	if changed {
		c.changes++
	}
}

func (c *codexCache) flush() {
	if d := len(c.next) - len(c.stored); len(c.stored) > 0 && c.changes <= 1 && d >= -1 && d <= 1 {
		return
	}
	var b bytes.Buffer
	fmt.Fprintf(&b, `{"version":%d,"sessions":{`, codexCacheVersion)
	for i, p := range c.order {
		e := c.next[p]
		if i > 0 {
			b.WriteByte(',')
		}
		pyQuote(&b, p)
		fmt.Fprintf(&b, `:{"mtime_ns":%d,"size":%d,"cwd":`, e.MtimeNS, e.Size)
		pyQuote(&b, e.Cwd)
		b.WriteString(`,"session_id":`)
		pyQuote(&b, e.SessionID)
		b.WriteString(`,"title":`)
		if e.Title == nil {
			b.WriteString("null")
		} else {
			pyQuote(&b, *e.Title)
		}
		fmt.Fprintf(&b, `,"title_scanned":%t}`, e.TitleScanned)
	}
	b.WriteString("}}")

	dir, name := "", c.path
	if i := strings.LastIndexByte(c.path, '/'); i >= 0 {
		dir, name = c.path[:i+1], c.path[i+1:]
	}
	if dir != "" && os.MkdirAll(dir, 0o777) != nil {
		return
	}
	tmp := dir + pyStem(name) + ".tmp" // Path.with_suffix(".tmp")
	if os.WriteFile(tmp, b.Bytes(), 0o666) != nil {
		return
	}
	_ = os.Rename(tmp, c.path)
}

// pyQuote writes s as json.dump does with ensure_ascii; undecodable bytes are
// written as the lone surrogates Python's surrogateescape would produce.
func pyQuote(b *bytes.Buffer, s string) {
	const hex = "0123456789abcdef"
	u := func(r rune) {
		b.WriteString(`\u`)
		b.WriteByte(hex[r>>12&0xf])
		b.WriteByte(hex[r>>8&0xf])
		b.WriteByte(hex[r>>4&0xf])
		b.WriteByte(hex[r&0xf])
	}
	b.WriteByte('"')
	for i := 0; i < len(s); {
		r, sz := utf8.DecodeRuneInString(s[i:])
		if r == utf8.RuneError && sz == 1 {
			u(0xdc00 | rune(s[i]))
			i++
			continue
		}
		i += sz
		switch r {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\n':
			b.WriteString(`\n`)
		case '\r':
			b.WriteString(`\r`)
		case '\t':
			b.WriteString(`\t`)
		case '\b':
			b.WriteString(`\b`)
		case '\f':
			b.WriteString(`\f`)
		default:
			switch {
			case r >= 0x20 && r <= 0x7e:
				b.WriteRune(r)
			case r > 0xffff:
				r -= 0x10000
				u(0xd800 | (r>>10)&0x3ff)
				u(0xdc00 | r&0x3ff)
			default:
				u(r)
			}
		}
	}
	b.WriteByte('"')
}
